### ISSUE TITLE
Proposal to restructure our watch channel cogs (big-brother, talent-pool)

### ISSUE THREAD

**Issue by SebastiaanZ** [python-discord/bot#317]
With the introduction of the helper nomination feature for the bot, our big-brother cog became much more complex and quite difficult to maintain. It implements similar, but slightly different features, which results in a lot of slightly different methods and/or conditional code within methods. That's why I like the idea of the initial implementation of @jchristgit in the Django branch: Have a separate cog for *Nomination* that inherits from *BigBrother*. In my opinion, this makes a lot of sense now the API endpoints have been separated anyway. However, I would like to propose a slightly different tactic for the inheritance, namely a *WatchChannel* base class both inherit from.

**WatchChannel base class**
In the current structure, with Nomination inheriting from BigBrother, changes to the BigBrother cog may be reflected in Nomination cog if it concerns a method that’s not overwritten in the child class. Since I don't agree with this hierarchy, I would like to propose a *WatchChannel* base class in bot/utils with the following inheritance diagram:

![Inheritance schematic](https://user-images.githubusercontent.com/33516116/53285591-905f5400-3762-11e9-84ff-d60eaf2b7797.png)

The base class should define the basic, but complete mechanisms for a watch channel, including the consumption, cache update, watch and unwatch features, in a **granulated** manner. The level of the granulation should depend on the needs to have different implementations for features in the different child classes. 

As an example, the header embeds for BigBrother and Nomination are different, but the rest of the header functionality is the same. By creating a separate method for header creation, we could easily just overwrite that part in our child classes, while keeping the rest of the code (that sends the header) consistent and DRY. Obviously, the granulation level should make sense, but this is an example to show the basic idea.

This is the gist of it and below is a list of the pros and cons I can think of.

**Pros**:

- Independence between BigBrother and Nomination: A change in one cog does not force a change in the inherited cog;
- DRY: we only have to overwrite that granules of code that are actually different;
- Separation of base functionality and implementation: We no longer have a massive, hybrid cog that implements two related, but ultimately different features (although the Django branch already has some separation with Nomination inheriting from BigBrother)

**Cons**:

- The abstraction separates the code over two files. However, this is already kinda the case with Nomination inheriting from BigBrother in the Django branch;
- Slightly more difficult code for beginning contributors to grasp.

Feel free to add your own thoughts to this thread or shoot my idea down completely.

**Comment by scragly** [comment]
As the two are intrinsically related through the base class, these should likely belong in the one extension:
```
watch_tools/
    __init__.py
    base.py
    big_brother.py
    nominations.py
```
```py
# __init__.py
from .big_brother import BigBrother
from .nominations import Nominations

def setup(bot):
    bot.add_cog(BigBrother)
    bot.add_cog(Nominations)
```
```py
# big_brother.py / nominations.py
from .base import WatchBase
```
```py
# bot/__main__.py
...
bot.load_extension("bot.cogs.watch_tools")
```

This ensures the extension can successfully reload during development, refreshing changes that have occurred in the base class which wouldn't occur if the base class was in a general util package. It also helps contain relevant models together, while still essentially doing the same thing as you're proposing.

**Comment by SebastiaanZ** [comment]
I really like this suggestion, so this is the way I would like to go.

**Comment by MarkKoz** [comment]
Implemented by #376
